Rewrite SSZ spec (#696)
ethereum/eth2.0-specs: specs/simple-serialize.md

# SimpleSerialiZe (SSZ)

This is a **work in progress** describing typing, serialization and Merkleization of Ethereum 2.0 objects.

## Table of contents

- [Typing](#typing)
    - [Basic types](#basic-types)
    - [Composite types](#composite-types)
    - [Aliases](#aliases)
- [Serialization](#serialization)
    - [`uintN`](#uintn)
    - [`bool`](#bool)
    - [Containers, tuples, lists](#containers-tuples-lists)
- [Deserialization](#deserialization)
- [Merkleization](#merkleization)
- [Self-signed containers](#self-signed-containers)
- [Implementations](#implementations)

## Typing

### Basic types

* `uintN`: `N`-bit unsigned integer (where `N in [8, 16, 32, 64, 128, 256]`)

> * `bool`: 1-bit unsigned integer

Author: Conceptually it's 1-bit. See [here](https://github.com/ethereum/eth2.0-specs/blob/bac352e70f825da9ac5a7155a5a8581033de15c3/specs/simple-serialize.md#bool) for the 1-byte serialisation.

> * `bool`: 1-bit unsigned integer

Reviewer: But in practice it's 1 byte.
Not sure if noting that it is "1-bit" is worthwhile at this point in the spec. A reader will naturally think (or at least ponder if) bools are serialized as 1 bit.

> * `bool`: 1-bit unsigned integer

Author: Changed it to

> * `bool`: `True` or `False`

### Composite types

* **container**: ordered heterogenous collection of values

>     * key-pair curly braket notation `{}`, e.g. `{'foo': uint64, 'bar': bool}`

Author: I see that [Wikipedia says](https://en.wikipedia.org/wiki/Bracket#Curly_brackets) it's "braces" in the US :joy:

>     * key-pair curly braket notation `{}`, e.g. `{'foo': uint64, 'bar': bool}`

Reviewer: lol. I just meant "braket" without the expected "c". "Bracket" I think is fine

>     * key-pair curly braket notation `{}`, e.g. `{'foo': uint64, 'bar': bool}`

Reviewer 2: Yeah, I'd say just use bracket, with modifiers (square, curly, angle) where needed, except "parentheses" for the round ones (ie. these).

* **tuple**: ordered fixed-length homogeneous collection of values
    * angle braket notation `[N]`, e.g. `uint64[N]`
* **list**: ordered variable-length homogenous collection of values
    * angle braket notation `[]`, e.g. `uint64[]`

### Aliases

For convenience we alias:

* `byte` to `uint8`
* `bytes` to `byte[]`
* `bytesN` to `byte[N]`

## Serialization

> We reccursively define the `serialize` function which consumes an object `value` (of the type specified) and returns a byte string of type `bytes`.

Author: > the language implementation might interpret it rather than defining the output as an SSZ type itself

That doesn't work great because we define `serialize` recursively, so the output of `serialize` itself needs to be an object of a well-defined type.

> We reccursively define the `serialize` function which consumes an object `value` (of the type specified) and returns a byte string of type `bytes`.

Reviewer: I see. sounds good

> #### `uintN`

Reviewer: reduce header level to `###`

```suggestion
### `uintN`
```

```python
assert N in [8, 16, 32, 64, 128, 256]
return value.to_bytes(N // 8, 'little')
```

> #### `bool`

Reviewer: ```suggestion
### `bool`
```

```python
assert value in (True, False)
return b'\x01' if value is True else b'\x00'
```

> #### Containers, tuples, lists

Reviewer: ```suggestion
### Containers, tuples, lists
```

```python

> serialized_bytes = ''.join([serialize(element) for element in value])

Reviewer: Note sure that I like the ssz type to be implicit in `element`. I think the more explicit `def serialize(value, ssz_type)` is clearer.

Reviewer 2: We already to `tree_hash_root(value)` all over the place with the type implicit; I feel like it's been clear enough for our purposes.

Author: Agreed with Vitalik :) I'm adding the following note as a compromise.

> *Note*: In the function definitions below (`serialize`, `hash_tree_root`, `signed_root`, etc.) objects implicitly carry their type.

> LENGTH_BYTES = 4

Reviewer: I think it clearer to keep the constant section at the top and define `LENGTH_BYTES` and `CHUNK_SIZE` there. And then in `Merkleization` section to call them "`CHUNK_SIZE`-byte chunks"

assert len(serialized_bytes) < 2**(8 * LENGTH_BYTES)
serialized_length = len(serialized_bytes).to_bytes(LENGTH_BYTES, 'little')
return serialized_length + serialized_bytes
```

## Deserialization

Given a type, serialization is an injective function from objects of that type to byte strings. That is, deserialization—the inverse function—is well-defined.

## Merkleization

We first define helper functions:

* `pack`: Given ordered objects of the same basic type, serialize them, pack them into 32-byte chunks, right-pad the last chunk with zero bytes, and return the chunks.

> * `merkleize`: Given ordered 32-byte chunks, right-pad them with zero chunks to the next power of two, Merkleize the chunks, and return the root.

Reviewer: Make it clearer that the number of chunks should be a power of two

Author: How do you suggest changing "right-pad them with zero chunks to the next power of two"?

Author: Reworded to "if necessary append zero chunks so that the number of chunks is a power of two" 👍

> * `mix_in_length`: Given a Merkle root `root` and a length `length` (32-byte little-endian serialization) return `hash(root + length)`.

Reviewer: ```suggestion
* `mix_in_length`: Given a Merkle root `root` and a length `length` (`uint256` little-endian serialization) return `hash(root + length)`.
```

* **container**: ordered heterogenous collection of values
    * key-pair curly braket notation `{}`, e.g. `{'foo': "uint64", 'bar': "bool"}`
* **tuple**: ordered fixed-length homogeneous collection of values
    * angle braket notation `[N]`, e.g. `uint64[N]`
* **list**: ordered variable-length homogenous collection of values
    * angle braket notation `[]`, e.g. `uint64[]`

### Aliases

For convenience we alias:

* `byte` to `uint8`
* `bytes` to `byte[]`
* `bytesN` to `byte[N]`

## Serialization

We recursively define the `serialize` function which consumes an object `value` (of the type specified) and returns a byte string of type `bytes`.

### `uintN`

```python
assert N in [8, 16, 32, 64, 128, 256]
return value.to_bytes(N // 8, 'little')
```

### `bool`

```python
assert value in (True, False)
return b'\x01' if value is True else b'\x00'
```

### Containers, tuples, lists

```python
serialized_bytes = ''.join([serialize(element) for element in value])
assert len(serialized_bytes) < 2**(8 * BYTES_PER_LENGTH_PREFIX)
serialized_length = len(serialized_bytes).to_bytes(BYTES_PER_LENGTH_PREFIX, 'little')
return serialized_length + serialized_bytes
```

## Deserialization

> Given a type, serialization is an injective function from objects of that type to byte strings. That is, deserialization—the inverse function—is well-defined.

Reviewer 2: This sounds confusing and not sure how many people will get what this means.

Maybe something like "Because serialization is an injective function (ie. if two objects of the same type are not exactly identical they will serialize to different values), for any bytestring there will always be at most one object of any given type that it could deserialize to (or zero if a bytestring is not valid SSZ). An efficient algorithm for computing this object can be found in the implementation [here](http://github.com/fill/in/the/correct/link)."

## Merkleization

We first define helper functions:

* `pack`: Given ordered objects of the same basic type, serialize them, pack them into `BYTES_PER_CHUNK`-byte chunks, right-pad the last chunk with zero bytes, and return the chunks.
* `merkleize`: Given ordered `BYTES_PER_CHUNK`-byte chunks, right-pad them with zero chunks to the next power of two, Merkleize the chunks, and return the root.
* `mix_in_length`: Given a Merkle root `root` and a length `length` (`uint256` little-endian serialization) return `hash(root + length)`.

We now define Merkleization `hash_tree_root(value)` of an object `value` recursively:

* `merkleize(pack(value))` if `value` is a basic object or a tuple of basic objects
* `mix_in_length(merkleize(pack(value)), len(value))` if `value` is a list of basic objects
* `merkleize([hash_tree_root(element) for element in value])` if `value` is a tuple of composite objects or a container

> * `mix_in_length(merkleize([hash_tree_root(element) for element in value]), len(value))` if `value` is a list of composite objects

Reviewer 2: > if `value` is a list of basic objects

I'd personally prefer "if `value` is a list of objects whose serialized values have a length of <32 bytes. The reason is that in the future we may want to serialize arrays of tuples of small variables and have them be packed efficiently; for one example of this see the [CBC proposal](https://github.com/ethereum/eth2.0-specs/issues/701) which includes 13-byte containers.

I would say consider something like this:

* `pack`: Given ordered objects of the same type, serialize them, right-pad each object with zero bytes to the next power of 2 if the size of the object is not a power of 2, pack them into `BYTES_PER_CHUNK`-byte chunks, right-pad the last chunk with zero bytes, and return the chunks.

We now define Merkleization `hash_tree_root(value)` of an object `value` recursively: 

* `merkleize(pack(value))` if `value` is an object or tuple of objects where each object has serialized size <= 32 bytes.
 * `mix_in_length(merkleize(pack(value)), len(value))` if `value` is a list of objects where each object has serialized size <= 32 bytes.
* `merkleize([hash_tree_root(element) for element in value])` if value is a tuple of objects with serialized size > 32 bytes or a container.
* `mix_in_length(merkleize([hash_tree_root(element) for element in value]), len(value))` if `value` is a list of objects with serialized size > 32 bytes.

Author: <img width="499" alt="screenshot 2019-02-28 at 11 28 25" src="https://user-images.githubusercontent.com/731743/53560127-fc1f3380-3b4b-11e9-89b2-d89442516a79.png">

# SimpleSerialiZe (SSZ)

This is a **work in progress** describing typing, serialization and Merkleization of Ethereum 2.0 objects.

## Table of contents

- [Constants](#constants)
- [Typing](#typing)
    - [Basic types](#basic-types)
    - [Composite types](#composite-types)
    - [Aliases](#aliases)
- [Serialization](#serialization)
    - [`uintN`](#uintn)
    - [`bool`](#bool)
    - [Containers, tuples, lists](#containers-tuples-lists)
- [Deserialization](#deserialization)
- [Merkleization](#merkleization)
- [Self-signed containers](#self-signed-containers)
- [Implementations](#implementations)

## Constants

| Name | Value | Description |
|-|-|-|
| `BYTES_PER_CHUNK` | `32` | Number of bytes per chunk.
| `BYTES_PER_LENGTH_PREFIX` | `4` | Number of bytes per serialized length prefix. |

## Typing
### Basic types

* `uintN`: `N`-bit unsigned integer (where `N in [8, 16, 32, 64, 128, 256]`)
* `bool`: `True` or `False`

### Composite types

* **container**: ordered heterogenous collection of values
    * key-pair curly bracket notation `{}`, e.g. `{'foo': "uint64", 'bar': "bool"}`

> * **tuple**: ordered fixed-length homogeneous collection of values

Reviewer 3: aren't these better called `array`:s? `tuples` tend to allow heterogeneous types

Author: I don't care either way. Implementations can use whatever terminology works best. For example, the natural terminology for Go is "array" (instead of tuple) and "slice" (instead of list). Notice that in Python lists also allow heterogeneous types 🤷‍♂️

# SimpleSerialiZe (SSZ)

This is a **work in progress** describing typing, serialization and Merkleization of Ethereum 2.0 objects.

## Table of contents

- [Constants](#constants)
- [Typing](#typing)
    - [Basic types](#basic-types)
    - [Composite types](#composite-types)
    - [Notation](#notation)
    - [Aliases](#aliases)
- [Serialization](#serialization)
    - [`uintN`](#uintn)
    - [`bool`](#bool)
    - [Containers](#containers)
    - [Tuples](#tuples)
    - [Lists](#lists)
- [Deserialization](#deserialization)
- [Merkleization](#merkleization)
- [Self-signed containers](#self-signed-containers)
- [Implementations](#implementations)

## Constants

| Name | Value | Description |
|-|-|-|
| `LENGTH_BYTES` | `4` | Number of bytes for the length of variable-length serialized objects. |

> | `MAX_LENGTH` | `2**(8 * LENGTH_BYTES)` | Maximum serialization length. |

Reviewer 4: Should the chunk size be defined here as well (with a note that it must be greater or equal than the maximum size of a basic type)

Author: The chunk size is chosen to match the output size of `hash` so is not really configurable.

* **container**: key-pair curly braket notation `{}` (e.g. `{'key1': uint64, 'key2': bool}`)
* **tuple**: angle braket notation `[N]` (e.g. `uint64[N]`)
* **list**: angle braket notation `[]` (e.g. `uint64[]`)

#### Aliases

For convenience we alias:

* `byte` to `uint8`
* `bytes` to `byte[]`
* `bytesN` to `byte[N]`
* `bit` to `bool`

## Serialization

We reccursively define the `serialize` function which consumes an object `object` (of the type specified) and returns a byte string of type `bytes`.

#### `uintN`

```python
assert N in [8, 16, 32, 64, 128, 256]
return object.to_bytes(N / 8, 'little')
```

#### `bool`

```python
assert object in (True, False)
return b'\x01' if object is True else b'\x00'
```

#### Containers

```python
serialized_elements = [serialize(element) for element in object]
serialized_bytes = reduce(lambda x, y: x + y, serialized_elements)
assert len(serialized_bytes) < MAX_LENGTH
serialized_length = len(serialized_bytes).to_bytes(LENGTH_BYTES, 'little')
return serialized_length + serialized_bytes
```

#### Tuples

```python
serialized_elements = [serialize(element) for element in object]
serialized_bytes = reduce(lambda x, y: x + y, serialized_elements)

> return serialized_bytes

Reviewer 4: Why don't tuples have a length prefix? I thought that maybe the reason is that it's not necessary because one knows the number of elements in advance, but the same is true for containers which do have the prefix.

Author: > the reason is that it's not necessary because one knows the number of elements in advance

Right! I think the idea is that developers can free up a buffer of the right size when reading data on the wire.

> the same is true for containers which do have the prefix

For containers `len(serialized_bytes)` is not always known ahead of time.

It's a bit awkward that length is handled in three different ways for contains, lists and tuples. Maybe we can just use `len(serialized_bytes)` everywhere.

Author: @jannikluhn: I've gone ahead and used the exact same serialisation logic for containers, lists and tuples. It's slightly less efficient but goes well with the "Simple" theme of SSZ.

cc @djrtwo, @vbuterin

Author: Thanks @jannikluhn, @djrtwo, @vbuterin for the reviews. All comments have been addressed :)

```

#### Lists

```python
serialized_elements = [serialize(element) for element in object]
serialized_bytes = reduce(lambda x, y: x + y, serialized_elements)
assert len(serialized_elements) < MAX_LENGTH
serialized_length = len(serialized_elements).to_bytes(LENGTH_BYTES, 'little')
return serialized_length + serialized_bytes
```

## Deserialization

Given a type, serialization is an injective function from objects of that type to byte strings. That is, deserialization—the inverse function—is well-defined.

## Merkleization

We first define helper functions:

* `pack`: Given ordered objects of the same basic type, serialize them, pack them into 32-byte chunks, right-pad the last chunk with zero bytes, and return the chunks.

> * `merkleize`: Given ordered 32-byte chunks, right-pad them with zero chunks to the closest power of two, Merkleize the chunks, and return the root.

Reviewer 4: ```suggestion
* `merkleize`: Given ordered 32-byte chunks, right-pad them with zero chunks to the next power of two, Merkleize the chunks, and return the root.
```

as the closest power of two might be smaller
